plugins/jenkins-backend/src/service/router.ts

```diff
@@ -0,0 +1,124 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorHandler } from '@backstage/backend-common';
+import express from 'express';
+import Router from 'express-promise-router';
+import { Logger } from 'winston';
+import { JenkinsInfoProvider } from './jenkinsInfoProvider';
+import { JenkinsApiImpl } from './jenkinsApi';
+
+export interface RouterOptions {
+  logger: Logger;
+  jenkinsInfoProvider: JenkinsInfoProvider;
+}
+
+export async function createRouter(
+  options: RouterOptions,
+): Promise<express.Router> {
+  const { jenkinsInfoProvider } = options;
+
+  const jenkinsApi = new JenkinsApiImpl();
+
+  const router = Router();
+  router.use(express.json());
+
+  router.get(
+    '/v1/entity/:namespace/:kind/:name/projects',
+    async (request, response) => {
+      const { namespace, kind, name } = request.params;
+      const branch = request.query.branch;
+      let branchStr: string | undefined;
+
+      if (branch === undefined) {
+        branchStr = undefined;
+      } else if (typeof branch === 'string') {
+        branchStr = branch;
+      } else {
+        // this was passed in as something weird -> 400
+        // https://evanhahn.com/gotchas-with-express-query-parsing-and-how-to-avoid-them/
+        response
+          .status(400)
+          .send('Something was unexpected about the branch queryString');
+
+        return;
+      }
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+      });
+      const projects = await jenkinsApi.getProjects(jenkinsInfo, branchStr);
+
+      response.json({
+        projects: projects,
+      });
+    },
+  );
+
+  router.get(
+    '/v1/entity/:namespace/:kind/:name/job/:jobName/:buildNumber',
+    async (request, response) => {
+      const { namespace, kind, name, jobName, buildNumber } = request.params;
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+        jobName,
+      });
+
+      const build = await jenkinsApi.getBuild(
+        jenkinsInfo,
+        jobName,
+        parseInt(buildNumber, 10),
+      );
+
+      response.json({
+        build: build,
+      });
+    },
+  );
+
+  router.post(
+    '/v1/entity/:namespace/:kind/:name/job/:jobName/:buildNumber::rebuild',
+    async (request, response) => {
+      const { namespace, kind, name, jobName } = request.params;
+
+      const jenkinsInfo = await jenkinsInfoProvider.getInstance({
+        entityRef: {
+          kind,
+          namespace,
+          name,
+        },
+        jobName,
+      });
+
+      await jenkinsApi.buildProject(jenkinsInfo, jobName);
+
+      // TODO: return the buildNumber which was started.
```

Comment:
it's actually available in the url above 🤔

Reply:
The new build gets allocated a new number. If I have builds 2, 3, and 4 If I say "rebuild 2" it will use the params etc. from 2 but the new build will be 5. The client doesn't know what the new build is but I haven't looked at the jenkins API closely enough to see where I get that info.

This might be a useless TODO as nothing needs the data yet and when we come to needing it, it will be obvious where it is needed so I've taken out the comment.